components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1836,25 +1837,90 @@ where
+        } else if !self.fsm.peer.is_leader() {
+            // The follower must in PreBecomeLeaderState. Warm up the entry cache
+            // if the low and high index are exactly the same as the warmup range.
+            let state = self.fsm.peer.pre_become_leader_state.as_ref().unwrap();
+            let entry_cache_warmup_range = state.entry_cache_warmup_range();
+            let is_entry_cache_warmup_timeout = state.is_entry_cache_warmup_timeout();
+
+            // Generally speaking, when the res.low is the same as the warmup
+            // range start, the fetch result is exactly used for warmup.
+            // As the low index of each async_fetch task is different.
+            // There should exist only one exception. A async fetch task
+            // with same low index is triggered before the warmup task.
+            //
+            // In other words, if this result is invalid for warmup, then
+            // the warmup operation is failed.
+            if entry_cache_warmup_range.0 != low {
+                self.fsm.peer.mut_store().clean_async_fetch_res(low);
+                return;
+            }
+            let valid_for_warmup = match &res.ents {
+                Ok(ents) => {
+                    let acture_high = ents.last().map(|e| e.index);
+                    let expected_high = entry_cache_warmup_range.1 - 1;
+                    let valid = acture_high == Some(expected_high);
+                    if !valid {
+                        warn!(
+                            "warm up the entry cache failed";
+                            "region_id" => self.fsm.region_id(),
+                            "peer_id" => self.fsm.peer_id(),
+                            "acture_high" => acture_high.unwrap_or(0),
+                            "expected_high" => expected_high,
+                        );
+                    }
+                    valid
+                }
+                Err(e) => {
+                    warn!(
+                        "warm up the entry cache failed";
+                        "region_id" => self.fsm.region_id(),
+                        "peer_id" => self.fsm.peer_id(),
+                        "err" => ?e,
+                    );
+                    false
+                }
+            };
+            // Use the result to warm up the entry cache if valid.
+            if valid_for_warmup {
+                self.fsm
+                    .peer
+                    .mut_store()
+                    .update_async_fetch_res(low, Some(res));
+                self.fsm
+                    .peer
+                    .mut_store()
+                    .on_warmup_range_fetched(entry_cache_warmup_range);
+                // Clean the async fetch result immediately if it is not used.
+                self.fsm.peer.mut_store().update_async_fetch_res(low, None);
+            }
+            // Need not to ack here if warmup is timeout.
+            if !is_entry_cache_warmup_timeout {
```

Comment:
I think if it's timeout, we should ack transfer leader anyway. Otherwise the transfer leader is greatly delayed?